### PR TITLE
Implement for-each loops over arrays

### LINKED ISSUE #375: Add for-each loop for array elements
Currently noir only supports a ranged for loop with integer start and end indices such as the following:
```rs
let start = 0;
let end = 5;
for i in start .. end {
    ...
}
```
(where `i`, `start`, and `end` must all be `const`). Expanding on this, we could also support a special version for arrays:
```rs
for elem in array {
    ...
}
```
Here neither `elem` nor `array` need be const since the actual indices would be known either way.

### Motivation

This feature would not increase noir's expressiveness since the same can always be achieved with indices and extracting array elements manually. It would increase ergonomics somewhat and decrease the barrier for those coming to noir from other languages like rust which allow for-each loops over arrays themselves. 

### Implementation strategy

The easiest way to implement this would likely be as syntactic sugar. For example, the above loop could be automatically expanded into:

```rs
for i in 0 .. std::array::len(array) {
    let elem = array[i];
    ...
}
```
Where `i` is a fresh variable.

### PR DESCRIPTION
**Pull request by jfecher**
# Related issue(s)

Resolves #375 

# Description

## Summary of changes

Implements for-each loops as described in #375 via syntactic sugar. 

~~Currently there is a rare bug that I may fix. In the expression `for e in array { ... }` we clone `array` internally causing it to be re-evaluated which could cause bugs if it mutates local variables. I'll add a commit soon to hoist a new let statement over the entire loop and wrap the loop in a block containing the let statement and the loop itself making the full desugared form:~~ Pushed a commit fixing the bug. The full desugaring for for-each loops is as follows:
```rs
{
    let fresh_ident1 = array;
    for fresh_ident2 in 0 .. std::array::len(fresh_ident1) {
        let e = fresh_ident1[fresh_ident2];
        ...
    }
}
```

## Test additions / changes

Adds a quick nested for-each loop example to `6_array`

### CODE REVIEW

File: crates/nargo/tests/test_data/6_array/proofs/p.proof [hunk 1]
```diff
@@ -0,0 +1 @@
...254e612041e956eadcdd30647908850acbb9336bb75f2e9418da4fe3a3ecb31211470d0ef0d6fc33c0039911c174cb137bbf005e685b5e4221af1513cc7d0c2c703b572c8ef9a7c91407abf5f6e56ca0937a0626b127040c19c4daa628bfcb0004b37675d37f12784a96a55404aa20c4534475cd13c9476512e042bfda07fad544f20a8db4c6499888cae01cbafceda5099eefe8e3e0a75221166a10203aeae8d6fa17f3bec644129d19d1faf7cde8b1aa3e32d8adfcf5b71110121fb880bc5dddda2fb093c3964d6aff6209dcc60bb04b504f4f57f6fa9429dae30435ba92eed01a5a6d36b4cddb858e30d28f15603aa6806b66f1ac4ea629ada86016980bf62b93590cafc09f1cb5bbc289d54f7f8062e6ca0db4ab45e91a377343ea9febd396c103f1e254a63dfea004634842380c164c1cf7fb5b05fb0305166e76b64a75d786f6ac9771a5ff58cf0bff51f5cf91ae1fcaf1cf26eefd2a764c3c7c27577dbc5604e88dbc4e1ec81f78942e619c71e731351c605caea405a55d4d50629b2940eee12c94cf50c35eb8da2f91adca631ddacef43f3d03c70c86d05a9c598e318e15397397f5ff3ce61a3f1651804fe9876ba07d2328f981186aa71921c48c5a68f30647df6bfcc1901972379c3bd2ce74903196fe3581bf0bbae42140a605f1b3f75bb84bafb6528859e0a00e088a033a117f73ba1cc90b174966f854cc0dc5c0e2a0e3fb9395f8f4c57c86d588d28802ab2b572efe45a226cc5b500f431cc8dccd6e664e1f8c443cba137e213af3c0c993c9c51d728cbe2250132e08617abac026e0ad902215a7e0e10be81fa7deeb7989b3f43fecbc8d
```
Comment (TomAFrench) [inline crates/nargo/tests/test_data/6_array/proofs/p.proof:1]: Looks like this was committed by accident.

**Review by TomAFrench — COMMENTED** [review]
Much cleaner than my efforts. Looks like you've committed a proof by accident though along with other changes to the `Verifier.toml`s

I've got some (simple) questions on the usage of `span` but it would probably just end up being spam on this PR. Are you in the Aztec discord?

File: crates/nargo/tests/test_data/6_array/proofs/p.proof (hunk 1 above)
Comment (jfecher) [reply crates/nargo/tests/test_data/6_array/proofs/p.proof:1]: Indeed

File: crates/nargo/tests/test_data/6_array/src/main.nr [hunk 2]
```diff
@@ -1,48 +1,55 @@
...
+    z=y[4];
+    for i in 0..3 {
+       z  = z + x[i]*y[i];
+       for j in 0..2 {
+          z = z + x[i+j]-y[i+j];
        };
-   };
-   constrain (z==41472);
+    };
+    constrain (z ==11539);
```
Comment (jfecher) [inline crates/nargo/tests/test_data/6_array/src/main.nr:43]: (github is confused about this line being changed, it was not)

File: crates/nargo/tests/test_data/main_return/Verifier.toml [hunk 3]
```diff
@@ -1 +1 @@
-return = "8"
+return = "0x0000000000000000000000000000000000000000000000000000000000000008"
```
Comment (jfecher) [inline crates/nargo/tests/test_data/main_return/Verifier.toml:1]: Looks like running the tests locally now mangles the Verifier files

Comment (guipublic) [reply crates/nargo/tests/test_data/main_return/Verifier.toml]: we should remove verifier.toml from source control, they are generated during prove.

**Comment by guipublic** [comment]
Why do you copy the array with  `let fresh_ident1 = array;` and do not use directly `array` in the loop?

**Comment by jfecher** [comment]
> Why do you copy the array with `let fresh_ident1 = array;` and do not use directly `array` in the loop?

I mentioned this in the crossed-out bug that was fixed. if the `array` expression contains side-effects then cloning it multiple times would re-evaluate the side-effects. Consider the code:
```rs
let mut i = 0;
for e in { i += 1; [1, 2, 3] } {
    ... i unchanged in loop ...
}
i
```
Users would expect `i = 1` at the end but if array is cloned and used in each loop iteration i would be `4` (1 initial evaluation + 1 evaluation per loop iteration)

**Comment by guipublic** [comment]
In you example, you have an expression, so in that case creating a fresh variable to store the result is correct because the expression returns an array by-copy.
However, if the expression is already an array, we should use it by-reference, i.e not do a copy. Otherwise, code that modifies the array in the loop will not work:
```
let mut a = [1,2,3];
for e in a {
    a[0] = e+1;
}
```

**Comment by guipublic** [comment]
In your example, you have an expression, so in that case creating a fresh variable to store the result is correct because the expression returns an array by-copy.
However, if the expression is already an array, we should use it directly and not do a copy. Otherwise, code that modifies the array in the loop will not work:

```
let mut a = [1,2,3];
let mut c=0;
for e in a {
 a[1] = e;
 c=c+e;
}
```
c is 6 with a fresh variable and 5 without

**Comment by jfecher** [comment]
I suppose this would depend on what users expect to happen here since it is disallowed in rust. If we want to allow it should we say that using it in a for loop copies the array or that you can mutate the array as needed. If its the later we could run into issues when the length of arrays can be changed once we add in vectors - that we will presumably also want to loop over.

**Comment by jfecher** [comment]
Alternatively if we had a way to create intermediate bindings in our Ast without incurring a copy this translation would be correct. Perhaps we should add an explicit copy node.

**Comment by guipublic** [comment]
I tried in rust and 
```
for e in a {
 a[1] = e;
 c=c+e;
}
```
gives 6, while 
```
for i in 0..3 {
 a[1] = a[i];
 c=c+a[i];
}
```
gives 5, so it seems we match rust behavior by doing the copy, which is fine for me!

**Comment by jfecher** [comment]
Ah, good catch. I was thinking of holding multiple mutable references to an array where I'm assuming yours is copied since it has a known size

**Review by guipublic — APPROVED** [review]

**Pull request merged**